libparsec/crates/client/src/certif/device.rs

```diff
@@ -0,0 +1,178 @@
+// Parsec Cloud (https://parsec.cloud) Copyright (c) BUSL-1.1 2016-present Scille SAS
+
+use std::sync::Arc;
+
+use libparsec_client_connection::ConnectionError;
+use libparsec_platform_storage::certificates::GetCertificateError;
+use libparsec_protocol::authenticated_cmds::v4::device_create;
+use libparsec_types::{
+    anyhow, thiserror, Bytes, CertificateSignerOwned, DateTime, DeviceCertificate, LocalDevice,
+    MaybeRedacted, SigningKeyAlgorithm, UserID,
+};
+
+use crate::EventTooMuchDriftWithServerClock;
+
+use super::{
+    encrypt::CertifEncryptForUserError, greater_timestamp, CertificateBasedActionOutcome,
+    CertificateOps, GreaterTimestampOffset, InvalidCertificateError,
+};
+
+pub(crate) async fn create_new_device(
+    certificate_ops: &CertificateOps,
+    new_device: LocalDevice,
+    author: Arc<LocalDevice>,
+) -> Result<CertificateBasedActionOutcome, CertifDeviceError> {
+    // Loop is needed to deal with server requiring greater timestamp
+    let mut timestamp = certificate_ops.device.now();
+
+    loop {
+        let outcome = internal_create_new_device(
+            certificate_ops,
+            new_device.clone(),
+            author.clone(),
+            timestamp,
+        )
+        .await?;
+
+        match outcome {
+            DeviceInternalsOutcome::Done(outcome) => return Ok(outcome),
+            DeviceInternalsOutcome::RequireGreaterTimestamp(strictly_greater_than) => {
+                // TODO: handle `strictly_greater_than` out of the client ballpark by
+                // returning an error
+                timestamp = greater_timestamp(
+                    &certificate_ops.device.time_provider,
+                    GreaterTimestampOffset::User,
+                    strictly_greater_than,
+                );
+            }
+        }
+    }
+}
+
+// returns (device_certificate, redacted_device_certificate)
+pub(crate) fn generate_new_device(
+    new_device: LocalDevice,
```

Comment:
```suggestion
    new_device: &LocalDevice,
```

Could be a ref, that would limit the requirement of multiple clone